pygmt/clib/conversion.py

```diff
@@ -102,7 +104,19 @@ def dataarray_to_matrix(grid):
             )
         region.extend([coord.min(), coord.max()])
         inc.append(coord_inc)
-    matrix = as_c_contiguous(grid.values[::-1])
+    if inc[1] < 0:
```

Comment:
So this uses xarray's builtin [sort](https://xarray-test.readthedocs.io/en/latest/reshaping.html#sort) mechanism that simplifies the code by heaps.
```suggestion

    if any([i < 0 for i in inc]):  # Sort grid properly if there are negative increments
        inc = [abs(i) for i in inc]
        grid = grid.sortby(variables=list(grid.dims), ascending=True)

    matrix = as_c_contiguous(grid.values[::-1])
```
Note :warning:: **do not** commit this change directly, you'll need to manually edit the file and make the change to the lines accordingly.

Also, it would be good to see some unit tests to make sure it actually works for every possible combination of flipped latitude/longitude coordinates.

Reply:
That works. I suspect that the sorting operation is slower than simply flipping the array, but compared to everything involved with the projection routines, this is certainly not a limiting operation.

As for the tests, I'd be happy to try, but I'm not sure what the point would be. It seems that the test would be to check if xarray works properly, not pygmt. What kind of test were you thinking of?

**Notes**
* It appears that this works even if the DataSet doesn't supply coordinates (the input grid is unmodified).
* The only problem I could envision is if the user defined the longitude coordinates as `[0, 90, 180, 270, 0]`. In this case the sort operation would provide incorrect results. It is hard to imagine someone doing this. If there is a test for this scenario, I would probably redefine the last coordinate somewhere else in the code. As part of such a test, you might also want to check that the two data values are the same (I think that gmt might already do this).

Reply:
If speed is the concern, it would be best to pass in a file anyway. You're right that the test is more of a check on xarray than pygmt, but we're actually not testing that `dataarray_to_matrix` properly now so it's a good time as any to add it.

Reply:
> If speed is the concern, it would be best to pass in a file anyway. 

I'm not sure even GMT handles this properly. So passing in a file probably doesn't help much.